src/index.ts

```diff
@@ -1,69 +1,56 @@
 import ejs from "ejs";
-import { getContext, helpers, Property } from "@dylibso/xtp-bindgen";
+import { ArrayType, EnumType, getContext, helpers, MapType, ObjectType, Property, XtpNormalizedType } from "@dylibso/xtp-bindgen";
 
-function toRustType(property: Property): string {
-  if (property.$ref) return `types::${helpers.capitalize(property.$ref.name)}`;
-  switch (property.type) {
-    case "string":
-      if (property.format === "date-time") {
-        return "chrono::DateTime<chrono::Utc>";
-      }
-      return "String";
-    case "number":
-      if (property.format === "float") {
-        return "f32";
-      }
-      if (property.format === "double") {
-        return "f64";
+function toRustTypeX(type: XtpNormalizedType): string {
+  // turn into reference pointer if needed
+  const optionalize = (t: string) => {
+    return type.nullable ? `Option<${t}>` : t
+  }
+
+  switch (type.kind) {
+    case 'string':
+      return optionalize('String')
+    case 'int32':
+      return optionalize('i32')
+    case 'float':
+      return optionalize('f32')
+    case 'double':
+      return optionalize('f64')
+    case 'byte':
+      return optionalize('byte')
+    case 'date-time':
+      return optionalize("chrono::DateTime<chrono::Utc>")
+    case 'boolean':
+      return optionalize('bool')
+    case 'array':
+      const arrayType = type as ArrayType
+      return optionalize(`Vec<${toRustTypeX(arrayType.elementType)}>`)
+    case 'buffer':
+      return optionalize('Vec<u8>')
+    case 'object':
+      const oType = (type as ObjectType)
+      if (oType.properties?.length > 0) {
+        return optionalize(`types::${helpers.capitalize(oType.name)}`)
+      } else {
+        // we're just exposing the serde values directly for backwards compat
+        return optionalize("std::collections::HashMap<String, serde_json::Value>")
       }
-      return "i64";
```

Comment:
note: we have a type for i64 but don't yet support it because of json limitations. i have an issue filed for this.